`UnneededPercentX` accepts \n in `%x`-expressions

bquorning opened the pull request:
Fix #1397 - allow multiline `%x`-expressions in `UnneededPercentX`, so it doesn’t suggest using backticks instead.

bbatsov commented:
I think this should be a config option.

bbatsov commented on lib/rubocop/cop/style/unneeded_percent_x.rb:
```diff
@@ -8,7 +8,9 @@ class UnneededPercentX < Cop
 
         def on_xstr(node)
-          add_offense(node, :expression) if node.loc.expression.source !~ /`/
+          return if node.loc.expression.source =~ /[`\n]/
```
I'd rather count the lines of the source to avoid an embedded `\n` in the backquoted expression.

When doing code analysis one of the rules of thumb is to avoid using regexps unless totally necessary.

bbatsov commented:
While you're at it - you might also fix https://github.com/bbatsov/rubocop/issues/1020 (pretty much the same issue for regexp literals).

bbatsov commented:
> I think this should be a config option.

There are some nuances to this, though - some people might like to enforce `%x` for multi-line literals.

bquorning commented:
> > I think this should be a config option.

> There are some nuances to this, though - some people might like to enforce `%x` for multi-line literals.

But we don't add configuration options for ignoring parts of a cop, right? A cop should only be configurable so it enforces one style or another.

I don't think I would ever use backticks around a multi-line literal.

bquorning commented:
This turned out to be a lot of changes and a lot of tests – including a few pending auto-correct tests.

I think I need some feedback now :-)

bbatsov commented:
> But we don't add configuration options for ignoring parts of a cop, right? A cop should only be configurable so it enforces one style or another.

I can't really understand what you mean.

bbatsov commented on config/default.yml:
```diff
@@ -553,6 +553,10 @@ Style/TrivialAccessors:
 
+Style/UnneededPercentX:
+  # Enforce using %x when the command string spans multiple lines.
+  EnforcePercentXForMultiLine: false
```
Ah, I see what you mean - if this was `AllowPercentXForMultiline` it would probably be better. Or we can rename the cop and implement distinct styles which might be the best option. As it stands the name of the cop is kind of limiting.

bquorning replied:
Exactly. Maybe the cop could be renamed to `CommandLiteral` (inspired by `RegexpLiteral`)?

Then we keep enforcing backticks when a one-line literal doesn’t contain “inner backticks”, and enforce (or allow) `%w` around multi-line command literals.

Or should the configuration rather be `PreferredStyleForMultiline` with values `percent_x` and `backtick`, so that one of the styles is always enforced?

bbatsov replied:
> Exactly. Maybe the cop could be renamed to CommandLiteral (inspired by RegexpLiteral)?

Fine by me.



> Or should the configuration rather be PreferredStyleForMultiline with values percent_x and backtick, so that one of the styles is always enforced?

Not sure. And some people might like to use `%x` always...

bquorning replied:
I can think of 4 different “preferred style” configurations for this “CommandLiteral” cop:

 1. `backticks_unless_has_inner_backtick`: doesn't care whether the literal is single or multi-line.
 2. `always_backticks`: doesn't care about inner backticks.
 3. `always_percent_x`: doesn't care about inner backticks.
 4. `percent_x_for_multiline`: doesn't care about the style of single-line command literals.

And I’m not sure whether I’m ok with the “don’t care”’s for 1) and 4). I might like an option where multi-lines are always `%x`, and single-lines are backtick’ed except if there’s an inner backtick.

Would it make sense to configure the styles separately for single-line and multi-line literals?

bbatsov replied:
Hmm, this is getting pretty complex. :-)

Style 2 & 4 doesn't make much sense IMO. Guess we can have just two styles - `backticks`, `percent_x` and `mixed` (backticks on a single-line, %x otherwise). Or we can have just 2 styles and a couple of other variables that affect the behaviour of the `backticks` style - this might make more sense to the users and it will spare us from thinking of odd style names.

bquorning replied:
Sorry for abandoning this PR for the better part of 3 weeks :-/

I think the 3-style approach (`backticks`, `percent_x`, `mixed`) is the way to go. Then we just need a a configuration option for what to do when a backtick’ed command literal contains a backtick: Allow it, or change to use `%x`?

I was thinking (since @bbatsov mentioned #1020) that maybe the configuration option should be `MaxBackticks`, and work in the same way as `MaxSlashes` does for `Style/RegexpLiteral`?

Or maybe `Style/RegexpLiteral` should have `MaxSlashes` removed and instead have an option to allow or disallow slashes in a shash’ed regex?

Either way, I think it makes sense to have these two cops to work, and be configured, similarly.

bquorning replied:
Ping @bbatsov.

bbatsov replied:
> I think the 3-style approach (backticks, percent_x, mixed) is the way to go. Then we just need a a configuration option for what to do when a backtick’ed command literal contains a backtick: Allow it, or change to use %x?

Sounds reasonable.

> I was thinking (since @bbatsov mentioned #1020) that maybe the configuration option should be MaxBackticks, and work in the same way as MaxSlashes does for Style/RegexpLiteral?
> 
> Or maybe Style/RegexpLiteral should have MaxSlashes removed and instead have an option to allow or disallow slashes in a shash’ed regex?

I think MaxSlashes was a mistake - you either allow them or don't.

bbatsov commented:
So, is this good to go? I see you updated the PR, but I don't see configurable styles.

bquorning commented:
No, this is still a work in progress (I'll update the PR header).

bquorning commented:
There we go – ready for review.

Two questions:

 1. Should we rename the cop?
 2. Will `on_xstr` ever be called with anything *but* a “backtick literal” or a “percent_x literal”? In other words, can I change my `if-elsif` to an `if-else`?

bbatsov commented:
> Should we rename the cop?

Yes. How about `CommandLiteral`?

> Will on_xstr ever be called with anything but a “backtick literal” or a “percent_x literal”? In other words, can I change my if-elsif to an if-else?

Guess you can change this.

bquorning commented:
> Will on_xstr ever be called with anything but a “backtick literal” or a “percent_x literal”? In other words, can I change my if-elsif to an if-else?
> Guess you can change this.

I just found this https://github.com/whitequark/parser/blob/v2.2.0.3/lib/parser/lexer/literal.rb#L32-L34 – apparently there’s a “heredoc backtick” as well. I suppose that wouldn’t work with this cop, in its current state.

bquorning commented on lib/rubocop/cop/style/command_literal.rb:
```diff
@@ -0,0 +1,94 @@
+          if backtick_literal?(node)
+            replacement = %w(%x( ))
+          elsif percent_x_literal?(node)
+            replacement = %w(` `)
```
It occurs to me that I should take into account the configuration of `Style/PercentLiteralDelimiters`/`PreferredDelimiters` here. But how do I use another cop’s configuration? And how do I set that config in the tests?

bquorning commented on lib/rubocop/cop/style/command_literal.rb:
```diff
@@ -0,0 +1,94 @@
+          fail CorrectionNotPossible if contains_backtick?(node)
+
+          if backtick_literal?(node)
+            replacement = %w(%x( ))
```
It occurs to me that I should take into account the configuration of `Style/PercentLiteralDelimiters`/`PreferredDelimiters` here. But how do I use another cop’s configuration? And how do I set that config in the tests?

bbatsov replied:
You can make an instance of the cop and use its `config` method. As for the tests - you always stub this.

bquorning replied:
Thanks. I ended up mimicking the test setup for spec/rubocop/cop/style/indent_hash_spec.rb

bquorning commented on lib/rubocop/cop/style/command_literal.rb:
```diff
@@ -0,0 +1,94 @@
+        end
+
+        def percent_x_literal?(node)
+          node.loc.begin.source == '%x(' && node.loc.end.source == ')'
```
I will remove the `(` and `)`, since users can choose to use virtually any character.

bquorning commented:
The cop now explicitly ignores “heredoc backtick” command literals. I can implement support for those in a later PR, if there’s interest for it.

Otherwise, this PR is ready for review.

bquorning commented on config/default.yml:
```diff
@@ -212,6 +212,20 @@ Style/CollectionMethods:
 
+# Use ` or %x around command literals.
+Style/CommandLiteral:
+  EnforcedStyle: mixed
```
According to the style guide, this should probably be changed to `backticks`, right?

bbatsov replied:
Right. I might revisit those rules in the guide, though. Rarely had to write multi-line command literals, so I didn't really dwell on this much. :-)

bbatsov commented on lib/rubocop/cop/style/command_literal.rb:
```diff
@@ -0,0 +1,98 @@
+module RuboCop
+  module Cop
+    module Style
+      # This cop enforces using `` or %x around command literals.
```
Add a few `@example`s.

bbatsov commented on lib/rubocop/cop/style/command_literal.rb:
```diff
@@ -0,0 +1,98 @@
+      class CommandLiteral < Cop
+        include ConfigurableEnforcedStyle
+
+        MSG_USE_BACKTICKS = 'Prefer using backticks around command string.'
```
prefer using -> use

One of those days we should make all the messages uniform.

bbatsov commented:
Apart from my small remarks the code looks good.

bquorning commented:
Changed the default setting to be `backticks`, added a few examples and changed the message texts as suggested.

One question: This cop explicitly doesn’t work with heredoc commands. I tried adding support for it (and may complete it in another PR), but it does complicate the code a whole lot. Is it ok to leave it out for now?

bquorning commented:
I really should run the tests before pushing :flushed:

bbatsov commented:
That's OK for now. Next stop - the regexp cop. :-)

bquorning commented:
Sure, feel welcome to assign #1020 to me.
